### PR TITLE
Sort tied results by earliest date achieved first

### LINKED ISSUE #7605: Sort tied results by earliest date achieved first
Source email thread: "Tie results sorting"

WRC and WRT are both happy with the proposed change. 

> Ordering by date of achievement is indeed less arbitrary than ordering by name, in particular as you could always argue that achieving a certain results earlier deserves slightly more merit than "just" tying the same result later.

**The change:**
* Currently, tied results are sorted alphabetically. We want this to be changed so that whoever achieved the received on the earlier date would be displayed first. 
* For identical results achieved on the same day, defaulting back to alphabetical sorting seems reasonable.

_Note: Ensure that MBLD sorting prioritises points, then time, then lower number of cubes unsolved, THEN date achieved._

### PR DESCRIPTION
**Pull request by simonkellly**
Fixes #7605

### CODE REVIEW

**Comment by DanielEgdal** [comment]
Potentially, it also makes sense to change the way limits are done? For example, if several people are tied for 100th, I think all should be displayed instead of just one.

**Comment by simonkellly** [comment]
I think if its ordered properly (as in first achieved is higher in the list), then it's okay to cut off the list once it hits 100, even if they have the same result (as they are not in the top 100 persons). Not sure how WRT / others feel about this though.

**Comment by FinnIckler** [comment]
Do we want to ask WRT first? There is already an email thread going

**Comment by simonkellly** [comment]
Sure, I will email now

**Comment by FinnIckler** [comment]
Seems to be the consensus of WRT that it should be top 100 ranks

**Comment by simonkellly** [comment]
I haven't tested this commit locally yet, as I am currently updating my local db

**Comment by simonkellly** [comment]
Tested it and it is working as expected

**Comment by FinnIckler** [comment]
Doing some performance tests

**Comment by FinnIckler** [comment]
I am getting an `Mysql2::Error: Duplicate column name 'id'` error on `/results/rankings/333/single?show=100+results` when I cloned this. Is it just me?

**Comment by simonkellly** [comment]
It worked on my pc and the tests passed so I'm not sure about what that error is caused by. I did some brief performance tests and it was very similar for simpler queries, and about 50% slower with all filters turned on (not exact numbers).

**Comment by simonkellly** [comment]
I have reproduced that error 😎

File: WcaOnRails/app/controllers/results_controller.rb [hunk 1]
```diff
@@ -29,49 +29,82 @@ def rankings
     @show = splitted_show_param[0].to_i
     @is_persons = splitted_show_param[1] == "persons"
     @is_results = splitted_show_param[1] == "results"
-    limit_condition = "LIMIT #{@show}"
+    expanded_limit_condition = "LIMIT #{@show * 2}"
+
+    # Required fields gotten from LightResult
+    result_select = "
```
Comment (FinnIckler) [inline WcaOnRails/app/controllers/results_controller.rb:35]: I generally like that we aren't using result.* anymore, do you know what happens with the duplicate values that are generated here?

Comment (simonkellly) [reply WcaOnRails/app/controllers/results_controller.rb:35]: Duplicate values?

**Comment by simonkellly** [comment]
BTW this is still not working, but I wasn't able to properly test it while writing the code. I also want to remove some of the code duplication within the different queries, as it makes it really annoying to edit

File: WcaOnRails/app/controllers/results_controller.rb [hunk 2]
```diff
@@ -29,49 +29,82 @@ def rankings
     @show = splitted_show_param[0].to_i
     @is_persons = splitted_show_param[1] == "persons"
     @is_results = splitted_show_param[1] == "results"
-    limit_condition = "LIMIT #{@show}"
+    expanded_limit_condition = "LIMIT #{@show * 2}"
```
Comment (FinnIckler) [inline WcaOnRails/app/controllers/results_controller.rb:32]: I guess this guarantees that everyone will be shown (except for weird edge cases where more than 2n people are tied?)

File: WcaOnRails/app/controllers/results_controller.rb [hunk 3]
```diff
@@ -89,42 +122,55 @@ def rankings
...
+              #{@years_condition_result}
+              #{@gender_condition}
+            GROUP BY result.countryId
+          ) record
+          JOIN Results result ON result.#{value} = recordValue AND result.countryId = recordCountryId
+          JOIN Competitions competition on competition.id = competitionId
           #{@gender_condition.present? ? "JOIN Persons persons ON result.personId = persons.id and persons.subId = 1" : ""}
           WHERE 1
             #{@event_condition}
-            #{@years_condition_result}
```
Comment (FinnIckler) [inline WcaOnRails/app/controllers/results_controller.rb:116]: is this just formatting or did you really replace `years_condition_result` with` years_condition_competition` ? If so, why?

**Comment by FinnIckler** [comment]
yes, generally a great idea rewriting this query...

File: WcaOnRails/app/controllers/results_controller.rb (hunk 3 above)
Comment (simonkellly) [reply WcaOnRails/app/controllers/results_controller.rb:116]: This is just the formatting of the pr, the current code on master uses both in the same way

File: WcaOnRails/app/controllers/results_controller.rb (hunk 2 above)
Comment (simonkellly) [reply WcaOnRails/app/controllers/results_controller.rb:32]: This value can likely be a lot smaller (maybe the limit + 25 or somthing), as it's highly unlikely to run into a case where even 25 people are tied for 100th

Comment (DanielEgdal) [reply WcaOnRails/app/controllers/results_controller.rb:32]: Fewest moves is probably going to be over 25 tied near the 100th spot at one point

**Comment by simonkellly** [comment]
Converting this to a draft as I will probably leave this for a while to work on my other pr, as this probably requires more than just simple changes especially when considering performance or how we might optimise how results work in the relativly near future

**Comment by gregorbg** [comment]
Closing this PR because after abandoning PHP, this is no longer a simple "fix" on top of the old tables.
We should rethink how we store results fundamentally. A dedicated project + proposal will follow.

**Pull request closed without merging**